File: 1206-33-1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1206-33-1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.java
class Solution {
    public String getHappyString(int n, int k) {
        StringBuilder ans = new StringBuilder();
        int total = 3 * (1 << (n - 1));
        if (total < k) return "";
        while (n > 0 && k > 0) {
            total = 1 << (n - 1);
            for (char ch = 'a'; ch <= 'c'; ch++) {
                if (ans.length() > 0 && ans.charAt(ans.length() - 1) == ch) continue;
                if (k > total) {
                    k -= total;
                } else {
                    ans.append(ch);
                    break;
                }
            }
            n--;
        }
        return ans.toString();
    }
}

File: 1206-33-1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1206-33-1415-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.js
var getHappyString = function(n, k) {
    let ans = [];
    let total = 3 * (1 << (n - 1));
    if (total < k) return "";
    while (n > 0 && k > 0) {
        total = 1 << (n - 1);
        for (let ch of ['a', 'b', 'c']) {
            if (ans.length > 0 && ans[ans.length - 1] === ch) continue;
            if (k > total) {
                k -= total;
            } else {
                ans.push(ch);
                break;
            }
        }
        n--;
    }
    return ans.join('');
};
